pkg/credentialprovider/external.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package credentialprovider
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"os/exec"
+	"sort"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/registrycredentials"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/registrycredentials/v1alpha1"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	registrycredentials.AddToScheme(scheme)
+	v1alpha1.AddToScheme(scheme)
+}
+
+type externalProviderKeyring struct {
+	// map of registry matcher to provider (each provider may have multiple)
+	providers map[string]ExternalCredentialProvider
+	// reverse sorted index of registries
+	index []string
+}
+
+func NewExternalProviderKeyring(configPath string) (DockerKeyring, error) {
```

Comment:
I would expect this to take a pre-validated `RegistryCredentialConfig` as input.  `ValidateKubeletConfiguration` should confirm that the config file is valid and from then on the code should be dealing with the struct, not a string file name.